### DISCUSSION TITLE
error while loading local file to jbrowse2

### DISCUSSION THREAD

**Discussion by Aditya-Upa** [GMOD/jbrowse-components#3894] [Q&A] [answered]
I want to embed Jbrowse2 into my nextjs app (with typescript disabled). 
I prepared the local files according to the instructions given in this [site  ](http://gmod.org/wiki/JBrowse_2_Tutorial_PAG_2022#A_note_about_preparing_indexed_GFF3_files
): 

First I tried loading the files using the 'require' function of JS :
 const fasta = require('./Phyag_NZFS3770.fna')
const fai = require('./Phyag_NZFS3770.fna.fai')
const gff = require('./Phyag_NZFS3770_genbank.sorted.gff.gz')
const tbi = require('./Phyag_NZFS3770_genbank.sorted.gff.gz.tbi')

I received an error like this :
./pages/components/genome_data/fna.fai_files/Phyag_NZFS3770.fna.fai
Module parse failed: Unexpected token (1:12)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
> LGTS01000001.1	127069	112	80	81
| LGTS01000002.1	106982	128882	80	81
| LGTS01000003.1	132400	237314	80	81

Thereafter I tried static loading by putting my genome files in the public directory of nextjs : 
  const assembly = [  {
    name: 'Phytophthora agathidicida',
    aliases: ['Phyag_NZFS3770'],
    sequence: {
      type: 'ReferenceSequenceTrack',
      trackId: 'GRCh38-ReferenceSequenceTrack',
      adapter: {
        type: 'BgzipFastaAdapter',
        fastaLocation: {
          uri: '10.0.0.231:3000/Phyag_NZFS3770.fna',
        },
        faiLocation: {
          uri: '10.0.0.231:3000/Phyag_NZFS3770.fna.fai',
        },
      },
    },
  },]
  
But again I received an error :
Server Error
Error: [mobx-state-tree] Error while converting `{"config":{"assembly":[{"name":"Phytophthora agathidicida","aliases":["Phyag_NZFS3770"],"sequence":{"type":"ReferenceSequenceTrack","trackId":"GRCh38-ReferenceSequenceTrack","adapter":{"type":"BgzipFastaAdapter","fastaLocation":{"uri":"10.0.0.231:3000/Phyag_NZFS3770.fna"......}]}]}}}` to `ReactLinearGenomeView`:
  
  Can you please suggest to me what changes I can make to my web application to load these files ? I can load them alright in the desktop version of jbrowse2. But i need to integrate jbrowse2 into the app because I need to dynamically visualise various genomic features using the genome browser.

**Comment by cmdcolin** [comment]
the public folder method likely is fine (importing the .fa file in code could work too probably if you had the webpack file loader but public folder should be ok)

But, the "error while converting" error is probably due to the gziLocation not being specified. See https://github.com/GMOD/jbrowse-react-app-nextjs-demo/blob/main/app/config.ts#L8 for example...that is also a working nextjs but points to remote.files, in your example can use public folder.

The bgzipfastaadapter requires a gzi file which is a gzip index. To generate the gzi file you can use "gunzip yourfile.fa.gz; bgzip -i yourfile.fa; samtools faidx yourfile.fa.gz" 

if in doubt, copying from versions of the config file that jbrowse desktop made could help (e.g. the .jbrowse in the desktop start screen is a config file that can be inspected in a text editor and adapted for web by replacing the localPath with uri)

I know the errors are a little opaque...it didn't just say to add a gziLocation... but we might be able to find some way to help with this

**Reply by Aditya-Upa** [reply]
so i changed my config file as you suggested, and tried static loading  : 

 assemblies: [
    {
      name: 'Phytophthora agathidicida',
      aliases: ['Phyag_NZFS3770'],
      sequence: {
        type: 'ReferenceSequenceTrack',
        trackId: 'GRCh38-ReferenceSequenceTrack',
        adapter: {
          type: 'BgzipFastaAdapter',
          fastaLocation: {
            uri: '10.0.0.231:3000/Phyag_NZFS3770.fna.gz',
          },
          faiLocation: {
            uri: '10.0.0.231:3000/Phyag_NZFS3770.fna.gz.fai',
          },
          gziLocation: {
            uri: '10.0.0.231:3000/Phyag_NZFS3770.fna.gz.gzi',
          },
        },
        
      },  },
  ],

Now while i am **not** getting an **explicit** error message; I am getting Error: **HTTP 404** fetching 10.0.0.231:3000/Phyag_NZFS3770.fna.gz.fai

I double-checked the path. Also if this page didn't exist I wouldn't be able to **download** this particular file hosted on my **public** directory from my browser. Thanks for all your help!

**Reply by cmdcolin** [reply]
Are you saying that you can download this page with your browser but JBrowse is still giving a 404?

**Reply by Aditya-Upa** [reply]
Exactly! And it's all files ..I tried statically loading the sample files which jbrowse2 has used ..human genome files..I got the same 404 error! I tried changing the file permissions to 755...still, same error

**Reply by cmdcolin** [reply]
if you can  e.g. take a screenshare video or something and share, might get a better idea. it would just be quite odd that visiting a page in the browser would work (e.g. download the file) and loading the same file from jbrowse would give 404 so i am suspecting e.g. a typo in the jbrowse config or something like that?

**Comment by Aditya-Upa** [comment]
sure thing! 

https://github.com/GMOD/jbrowse-components/assets/83017903/a1e768f7-681c-4436-8182-06855130ce54

**Reply by cmdcolin** [reply]
ah, I think that helps to explain it. JBrowse needs you to put the "http://" or "https://" prefix on the URI if it is an 'absolute path' like this! thanks for the detailed video recording, i hope that helps

**Comment by Aditya-Upa** [comment]
now i am getting another error : Error: Encountered invalid refName: "�"

https://github.com/GMOD/jbrowse-components/assets/83017903/a281e375-98a2-4030-86a9-9b3b6ace290e

**Reply by cmdcolin** [reply]
this may indicate an error with gzip configuration with either (a) the source file itself, in which case, re-preparing the bgzip fasta may help or (b) the server may be returning incorrect headers on the gzip data, which is harder to fix. if (b) I'd just suggest using regular IndexedFastaAdapter or TwoBitAdapter instead of BgzipIndexedFasta. you can set up regular IndexedFastaAdapter by running in your command prompt 

```
gunzip Phyag_NZFS3770.fna.gz
samtools faidx Phyag_NZFS3770.fna
```

And then making your config

```
assemblies: [
{
name: 'Phytophthora agathidicida',
aliases: ['Phyag_NZFS3770'],
sequence: {
type: 'ReferenceSequenceTrack',
trackId: 'GRCh38-ReferenceSequenceTrack',
adapter: {
type: 'IndexedFastaAdapter',
fastaLocation: {
uri: '10.0.0.231:3000/Phyag_NZFS3770.fna',
},
faiLocation: {
uri: '10.0.0.231:3000/Phyag_NZFS3770.fna.fai',
}
},
```

**Comment by Aditya-Upa** [comment]
Thanks for the tip! I had already prepared the fasta file like that. While the gff file is getting loaded alright..I am facing another error while loading the fna file: 
![jbrowse2_1](https://github.com/GMOD/jbrowse-components/assets/83017903/dae6b14f-47f3-465f-8cd0-83b3776c5b8d)
then I added the gzi portion too and got the following error : 
![jbrowse2](https://github.com/GMOD/jbrowse-components/assets/83017903/2e1596f2-0e3a-49b0-bf63-725f91c5fc41)

**Reply by cmdcolin** [reply]
can you paste your config? the regular IndexedFastaAdapter shouldnt have any gzip related error

**Comment by Aditya-Upa** [comment]
sure here is the config file :
  const config = {
    assemblies: [
      {
        name: 'Phytophthora_agathidicida',
        aliases: ['Phyag_NZFS3770'],
        sequence: {
          type: 'ReferenceSequenceTrack',
          trackId: 'GRCh38-ReferenceSequenceTrack',
          adapter: {
            type: 'BgzipFastaAdapter',
            fastaLocation: {
              uri: 'http://10.0.0.231:3000/Phyag_NZFS3770.fna',
  
            },
            faiLocation: {
            uri: 'http://10.0.0.231:3000/Phyag_NZFS3770.fna.fai',
            },
            // gziLocation: {
            //   uri: 'http://10.0.0.231:3000/Phyag_NZFS3770.fna.gz.gzi',
            // },
          },
          
        },
        // refNameAliases: {
        //   adapter: {
        //     type: 'RefNameAliasAdapter',
        //     location: {
        //       uri: 'https://s3.amazonaws.com/jbrowse.org/genomes/GRCh38/hg38_aliases.txt',
        //     },
        //   },
        // },
      },
    ],
  
    tracks: [
      {
        type: 'FeatureTrack',
        trackId: 'genes',
        name: 'NCBI RefSeq Genes',
        assemblyNames: ['Phytophthora_agathidicida'],
        category: ['Genes'],
        adapter: {
          type: 'Gff3TabixAdapter',
          gffGzLocation: {
            uri: 'http://10.0.0.231:3000/Phyag_NZFS3770.sorted.gff3.gz',
          },
          index: {
            location: {
              uri: 'http://10.0.0.231:3000/Phyag_NZFS3770.sorted.gff3.gz.tbi',
            },
          },
        },
      },
      {
        type: 'FeatureTrack',
        trackId: 'repeats_hg38',
        name: 'SSR file',
        assemblyNames: ['Phytophthora_agathidicida'],
        category: ['Annotation'],
        adapter: {
          type: 'BigBedAdapter',
          bigBedLocation: {
            uri: 'http://10.0.0.231:3000/Phyag_NZFS3770.bw',
            locationType: 'UriLocation',
          },
        },
      },
    ],
  defaultSession: {
    name: 'this session',
    margin: 0,
    views: [
      {
        id: 'linearGenomeView',
        minimized: false,
        type: 'LinearGenomeView',
        offsetPx: 191980240,
        bpPerPx: 0.1554251851851852,
        // displayedRegions: [
        //   {
        //     refName: '10',
        //     start: 0,
        //     end: 133797422,
        //     reversed: false,
        //     assemblyName: 'Phyag',
        //   },
        // ],
        hideHeader: false,
        hideHeaderOverview: false,
        hideNoTracksActive: false,
        trackSelectorType: 'hierarchical',
        trackLabels: 'overlapping',
        showCenterLine: false,
        showCytobandsSetting: true,
        showGridlines: true,
      },
    ],
  },
};

export default config

**Reply by cmdcolin** [reply, marked as answer]
Change type: 'BgzipFastaAdapter' to type: 'IndexedFastaAdapter', I think that will fix it

**Comment by Aditya-Upa** [comment]
That was such a silly mistake on my part! thank you so much !
